utils/int_types.go

```diff
@@ -0,0 +1,66 @@
+package utils
+
+import (
+	"errors"
+
+	"github.com/KyberNetwork/int256"
+	"github.com/holiman/uint256"
+)
+
+// define placeholders for these types, in case we need to customize them later
+// (for example to add boundary check...)
+
+type Uint256 = uint256.Int
+type Uint160 = uint256.Int
+type Uint128 = uint256.Int
+
+type Int256 = int256.Int
+type Int128 = int256.Int
+
+var (
+	ErrExceedMaxInt256 = errors.New("exceed max int256")
+	ErrOverflowUint128 = errors.New("overflow uint128")
+	ErrOverflowUint160 = errors.New("overflow uint160")
+
+	Uint128Max = uint256.MustFromHex("0xffffffffffffffffffffffffffffffff")
+	Uint160Max = uint256.MustFromHex("0xffffffffffffffffffffffffffffffffffffffff")
+)
+
+// https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/SafeCast.sol
+func ToInt256(value *Uint256, result *Int256) error {
+	// if value (interpreted as a two's complement signed number) is negative -> it must be larger than max int256
+	if value.Sign() < 0 {
+		return ErrExceedMaxInt256
+	}
+	var ba [32]byte
+	value.WriteToArray32(&ba)
+	result.SetBytes32(ba[:])
+	return nil
+}
+
+// https://github.com/Uniswap/v3-core/blob/main/contracts/libraries/SafeCast.sol
+func CheckToUint160(value *Uint256) error {
+	// we're using same type for Uint256 and Uint160, so use the original for now
+	if value.Cmp(Uint160Max) > 0 {
```

Comment:
do we need to check if `value` is negative?

Reply:
um it's uint256/160 so there's no negative value here (for example we can interpret `Uint160Max`(`2^160-1`) as `-1` in int160, but for uint160 it's still a valid positive value)